cli/internal/fs/package_json.go

```diff
@@ -59,12 +60,30 @@ func (r *Workspaces) UnmarshalJSON(data []byte) error {
 }
 
 // ReadPackageJSON returns a struct of package.json
-func ReadPackageJSON(path turbopath.AbsoluteSystemPath) (*PackageJSON, error) {
-	b, err := path.ReadFile()
+func ReadPackageJSON(repoRoot turbopath.AbsoluteSystemPath, pkgJSONFilePath turbopath.AbsoluteSystemPath) (*PackageJSON, error) {
+	b, err := pkgJSONFilePath.ReadFile()
 	if err != nil {
 		return nil, err
 	}
-	return UnmarshalPackageJSON(b)
+	pkg, err := UnmarshalPackageJSON(b)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// get relative path for package.json
+	relativePkgJSONPath, err := repoRoot.PathTo(pkgJSONFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	// assign PackageJSONPath and Dir
+	// TODO(mehulkar) These could go into WorkspaceInfos or some other Workspace construct
+	// rather than onto the fs.PackageJSON itself (which should just be the unmarshaled contents of package.json)
+	pkg.PackageJSONPath = turbopath.AnchoredSystemPathFromUpstream(relativePkgJSONPath)
+	pkg.Dir = turbopath.AnchoredSystemPathFromUpstream(filepath.Dir(relativePkgJSONPath))
```

Comment:
I don't love it either, but I also felt it was weird that the `fs.PackageJSON` struct sometimes has these fields and sometimes doesn't. Is the TODO I left here enough to indicate that we should come back to it? I can go the route of the functional options, but I think that will reach into quite a few files. 

Another option is to manually set this for the root workspace:

https://github.com/vercel/turbo/blob/f1fc4bdb74940d1c6eb3b1b0ee10308e7937045a/cli/internal/run/run.go#L158-L159

like it's being set for other workspaces:

https://github.com/vercel/turbo/blob/f1fc4bdb74940d1c6eb3b1b0ee10308e7937045a/cli/internal/context/context.go#L370-L371

The problem with this is then I would not touch anything in prune, and it could become confusing why the Root Workspace is different for the `run` and `prune` and other commands.

Reply:
I think we'd have to do it everywhere. I do want a use case to use that pattern, but at the moment I'd rather get this merged and get back to #2706 so we can keep 1.8 release on schedule